File: python/house-robber-iii.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    # memoized dfs
    # TC: O(n) where n is the number of nodes in the tree
    # SC: O(n)
    def rob(self, root: Optional[TreeNode]) -> int:
        def dfs(node, taken, mem):
            if node == None: return 0
            
            if (node, taken) in mem:
                return mem[(node, taken)]
            
            if taken:
                mem[(node, taken)] = dfs(node.left, False, mem) + dfs(node.right, False, mem)
            else:
                mem[(node, taken)] = max(
                    dfs(node.left, True, mem) + dfs(node.right, True, mem) + node.val,
                    dfs(node.left, False, mem) + dfs(node.right, False, mem)
                )
            
            return mem[(node, taken)]
        
        return dfs(root, False, dict())

File: go/house_robber_iii.go
package leetcode

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func rob(root *TreeNode) int {
	return dfs(root, false, make(map[Key]int))
}

func dfs(node *TreeNode, taken bool, mem map[Key]int) int {
	if node == nil {
		return 0
	}

	key := Key{node, taken}
	if _, ok := mem[key]; ok {
		return mem[key]
	}

	if taken {
		mem[key] = dfs(node.Left, false, mem) + dfs(node.Right, false, mem)
	} else {
		mem[key] = max(
			dfs(node.Left, true, mem)+dfs(node.Right, true, mem)+node.Val,
			dfs(node.Left, false, mem)+dfs(node.Right, false, mem),
		)
	}
	val, _ := mem[key]
	return val
}

func max(x, y int) int {
	if x > y {
		return x
	} else {
		return y
	}
}

type Key struct {
	node  *TreeNode
	taken bool
}
